pkg/storegateway/indexheader/snapshotter.go

```diff
@@ -0,0 +1,145 @@
+package indexheader
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/ulid"
+
+	"github.com/grafana/mimir/pkg/util/atomicfs"
+)
+
+const lazyLoadedHeadersListFileName = "lazy-loaded.json"
+
+type SnapshotterConfig struct {
+	Enabled bool
+
+	// Path stores where lazy loaded blocks will be tracked in a single file per tenant
+	Path   string
+	UserID string
+}
+
+// Snapshotter manages the snapshots of lazy loaded blocks.
+type Snapshotter struct {
+	logger log.Logger
+	conf   SnapshotterConfig
+
+	done chan struct{}
+}
+
+func NewSnapshotter(logger log.Logger, conf SnapshotterConfig) *Snapshotter {
+	return &Snapshotter{
+		logger: logger,
+		conf:   conf,
+		done:   make(chan struct{}),
+	}
+}
+
+type blocksLoader interface {
+	LoadedBlocks() map[ulid.ULID]int64
+}
+
+func (s *Snapshotter) Start(ctx context.Context, l blocksLoader) error {
+	if !s.conf.Enabled {
+		return nil
+	}
+
+	err := s.persistLoadedBlocks(l)
+	if err != nil {
+		return fmt.Errorf("persist initial list of lazy-loaded index headers: %w", err)
+	}
+
+	go func() {
+		tick := time.NewTicker(time.Minute)
+		defer tick.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-s.done:
+				return
```

Comment:
why both `ctx.Done()` and `s.done`?

Reply:
Here, context is bound to the upper service's lifetime, but `s.done` signals on the call to `Snapshotter.Stop`.

If I read the code in `storegateway` right, the upper service cleans up its dependencies, while it's still running (ref [code](https://github.com/grafana/mimir/blob/486c0338d8f5f66aeeeba074a1e78563e5c37a59/pkg/storegateway/bucket.go#L263-L265))

Reply:
ah, right. ctx.Done is triggered on store-gateway shutdown, and s.done is triggered when the tenant no longer shards to this store-gateway